### ISSUE TITLE
Project references: TS server does not find correct project

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#56959]
### 🔎 Search Terms

- project references
- tsconfig.json

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about project references.

### 💻 Code

Full reduced test case: https://github.com/OliverJAsh/ts-project-references-nested

Contents inlined below.

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "outDir": "./dist/"
  },
  "references": [
    { "path": "./demos/tsconfig.json" },
    { "path": "./app/tsconfig.json" }
  ]
}

```
`app/tsconfig.json`:
```json
{
  "compilerOptions": {
    "composite": true,
    "outDir": "../app-dist/"
  },
  "include": ["**/*"],
  "exclude": ["**/*-demos.*"]
}

```
`app/Component-demos.ts`:
```ts
// This errors.
import * as helpers from 'demos/helpers';

export const demo = () => {
    helpers;
}

```
`demos/tsconfig.json`:
```json
{
  "compilerOptions": {
    "composite": true,
    "rootDir": "../",
    "outDir": "../demos-dist/",
    "paths": {
      "demos/*": ["./*"]
    }
  },
  "include": [
    "**/*",
    "../app/**/*-demos.*"
  ],
  "references": [{ "path": "../app/tsconfig.json" }]
}

```
`demos/helpers.ts`:
```ts
export const foo = 1;

```

### 🙁 Actual behavior

When I run `tsc --build --verbose tsconfig.json`, there are no errors, but when I open my editor (VS Code) and open the file `app/Component-demos.ts`, I see errors:

<img width="721" alt="image" src="https://github.com/microsoft/TypeScript/assets/921609/4dbc8ced-4809-4d4c-876c-17fafddfeb2c">

### 🙂 Expected behavior

No errors inside my editor.

### Additional information about the issue

When `app/Component-demos.ts` is opened, I believe TypeScript traverses parent file directories searching for a `tsconfig.json` file, and it stops at the first file it finds which in this case is `app/tsconfig.json`. This TS project does not include the file `app/Component-demos.ts`, so I believe TypeScript should continue traversing parent file directories to search for another `tsconfig.json` that does include the file. In this case, we want TypeScript to discover the root `tsconfig.json`.

**Comment by OliverJAsh** [comment]
@sheetalkamat I can still reproduce this issue with TypeScript 5.5.3—I've pushed another commit to my reduced test case. Just to check, has this been released?

**Comment by sheetalkamat** [comment]
This fix isn’t in 5.5 but will be in 5.6 .. you can try it out using nightly in the mean time

**Comment by sheetalkamat** [comment]
This will be included in later release after we can improve perf on looking up in referenced projects
